src/lib/Imp.hs

```diff
@@ -267,11 +270,11 @@ destToAtomScalarAction fScalar dest = do
 
 destToAtom' :: (IVar -> ImpM Atom) -> [Var] -> Dest -> EmbedT ImpM Atom
 destToAtom' fScalar forVars dest = case dest of
-  DFor (Abs v d) -> do
+  DFor a@(Abs v _) -> do
     v' <- lift $ case v of
       (NoName :> t) -> freshVar ("idx" :> t)
```

Comment:
As discussed, I'm worried that the embed monad doesn't know that these variables are in scope. Using `buildLam` should take care of it.